File: docs/anklewrite/testing/markdown.py
from re import S


class markdownPattern:
    def __init__(self, pattern):
        self.pattern = pattern;

        self.loggedStartIndex = False
        self.textCharInEmpty = False # Keeps track that at least one text character was found in the !

        self.tempStartIndex = 0
        self.tempEndIndex = 0
        self.currentIndex = 0
        self.wordsInBlankSpace = []
        self.wordInBlankSpace = ""
        pass

    def checkNextCharacterValid(self, character, index):
        patternMatched = False # Defaults to failing the pattern, waits to see if a case is satisfied
        if self.pattern[self.currentIndex] == character: # If the character selected is the input character, satisfy
            # print("Found a character")
            self.currentIndex += 1
            patternMatched = True
        elif self.pattern[self.currentIndex] == '!': # If the parser is currently in blank space
            if self.pattern[self.currentIndex + 1] == character and self.textCharInEmpty: # Checks if the blank space is over, using textCharInEmpty to ensure that there was at least one character in the blank space
                self.currentIndex += 2
                self.textCharInEmpty = False
                patternMatched = True
                # print("Found a character")
                self.wordsInBlankSpace.append(self.wordInBlankSpace)
                self.wordInBlankSpace = ""
            elif self.pattern[self.currentIndex + 1] == character and not self.textCharInEmpty: # Opposite of above, if there was no blank space character, fail the check (ex "*i* vs **")
                # print("Found a character, but there was no character in the blank space (!)")
                patternMatched = False
            else: # For any other letter that doesn't match
                # print("Foound a blank space character")
                self.wordInBlankSpace += character
                self.textCharInEmpty = True
                patternMatched = True
        
        if self.currentIndex == 1 and not self.loggedStartIndex: # Log the start index if not already done
            self.tempStartIndex = index
            self.loggedStartIndex = True
        
        if len(self.pattern) == self.currentIndex: # If the currentIndex is at the end of the pattern, complete the pattern recognition
            self.tempEndIndex = index
            # print("Found pattern " + self.pattern + " starting at " + str(self.tempStartIndex) + " and ending at " + str(self.tempEndIndex))
            # print(str(self.wordsInBlankSpace) + " was contained in blank space")
            result = markdownResult(self.tempStartIndex, self.tempEndIndex, self.wordsInBlankSpace, self.pattern)
            self.reset()
            return result
            
        if not patternMatched and self.currentIndex > 0: # If no case was triggered above and the match failes
            # print("Matching failed, restarting")
            self.reset()
        pass

    def reset(self):
        self.currentIndex = 0
        self.textCharInEmpty = False
        self.tempEndIndex = 0
        self.loggedStartIndex = False
        self.wordInBlankSpace = ""
        self.wordsInBlankSpace = []

class markdownResult:
    def __init__(self, startIndex, endIndex, wordsInBlankSpace, pattern):
        self.startIndex = startIndex
        self.endIndex = endIndex
        self.wordsInBlankSpace = wordsInBlankSpace
        self.pattern = pattern

    def __str__(self):
        return "Found pattern " + self.pattern + " starting at " + str(self.startIndex) + " and ending at " + str(self.endIndex) + "\n" + str(self.wordsInBlankSpace) + " was contained in blank space"
        
patterns = [ markdownPattern("[!](!)"), markdownPattern("*!*"), markdownPattern("**!**") ]

markdownPatterns = []

inputString = "Markdown *is* cool, my **parser** can even handle links like [Test](https://www.google.com)."

for x in range(len(inputString)) :
    currentChar = inputString[x]
    for pattern in patterns:
        # print(str(x) + ": " + currentChar)
        rReturn = pattern.checkNextCharacterValid(currentChar, x)
        if rReturn:
            markdownPatterns.append(rReturn)

for pattern in markdownPatterns:
    print(pattern)
    print("\n")

File: docs/anklewrite/markdown.js
var currentChar, inputString, markdownPatterns, patterns, rReturn;

class markdownPattern {
  constructor(pattern) {
    this.pattern = pattern;
    this.loggedStartIndex = false;
    this.textCharInEmpty = false;
    this.tempStartIndex = 0;
    this.tempEndIndex = 0;
    this.currentIndex = 0;
    this.wordsInBlankSpace = [];
    this.wordInBlankSpace = "";
  }

  checkNextCharacterValid(character, index) {
    var patternMatched, result;
    patternMatched = false;

    if (this.pattern[this.currentIndex] === character) {
      this.currentIndex += 1;
      patternMatched = true;
    } else {
      if (this.pattern[this.currentIndex] === "!") {
        if (this.pattern[this.currentIndex + 1] === character && this.textCharInEmpty) {
          this.currentIndex += 2;
          this.textCharInEmpty = false;
          patternMatched = true;
          this.wordsInBlankSpace.push(this.wordInBlankSpace);
          this.wordInBlankSpace = "";
        } else {
          if (this.pattern[this.currentIndex + 1] === character && !this.textCharInEmpty) {
            patternMatched = false;
          } else {
            this.wordInBlankSpace += character;
            this.textCharInEmpty = true;
            patternMatched = true;
          }
        }
      }
    }

    if (this.currentIndex === 1 && !this.loggedStartIndex) {
      this.tempStartIndex = index;
      this.loggedStartIndex = true;
    }

    if (this.pattern.length === this.currentIndex) {
      this.tempEndIndex = index;
      result = new markdownResult(this.tempStartIndex, this.tempEndIndex, this.wordsInBlankSpace, this.pattern);
      this.reset();
      return result;
    }

    if (!patternMatched && this.currentIndex > 0) {
      this.reset();
    }
  }

  reset() {
    this.currentIndex = 0;
    this.textCharInEmpty = false;
    this.tempEndIndex = 0;
    this.loggedStartIndex = false;
    this.wordInBlankSpace = "";
    this.wordsInBlankSpace = [];
  }

}

class markdownResult {
  constructor(startIndex, endIndex, wordsInBlankSpace, pattern) {
    this.startIndex = startIndex;
    this.endIndex = endIndex;
    this.wordsInBlankSpace = wordsInBlankSpace;
    this.pattern = pattern;
  }

  shouldRenderChar(index, character) {
    // Check if the character is in the pattern
    if (index < this.startIndex || index > this.endIndex) {
      return true;
    }
    for (var i = 0; i < this.pattern.length; i++) {
      // console.log("testing against pattern, startindex is " + this.startIndex + " and endindex is " + this.endIndex + ", character " + character + " is at index " + index);
      if (this.pattern[i] == character) {
        // console.log("returned false")
        return false;

      }
    }

    if (this.pattern == "[!](!)") {
      // Don't render anything if it is a link
      return false;
    }
    return true;
  }
}


markdownPatternsToClass = {
  "*!*": "italics",
  "**!**": "bold",
  "[!](!)": "link"
}

function getMarkdownPatterns(input) {
    patterns = [ new markdownPattern("*!*"), new markdownPattern("**!**"), new markdownPattern("[!](!)")];
    markdownPatterns = [];

    for (var x = 0, _pj_a = input.length; x < _pj_a; x += 1) {
        currentChar = input[x];
      
        for (var pattern, _pj_d = 0, _pj_b = patterns, _pj_c = _pj_b.length; _pj_d < _pj_c; _pj_d += 1) {
          pattern = _pj_b[_pj_d];
          rReturn = pattern.checkNextCharacterValid(currentChar, x);
      
          if (rReturn) {
            markdownPatterns.push(rReturn);
          }
        }
      }
    return markdownPatterns;
}

function markdownToHtmlNotVisible(input) {
    var markdownPatterns = getMarkdownPatterns(input);
    var outputString = ""

    for (var x = 0; x < input.length; x++) {
        currentChar = input[x];
        for (var m = 0; m < markdownPatterns.length; m++) {
            if (markdownPatterns[m].startIndex === x + 1) {
                if (markdownPatternsToClass[markdownPatterns[m].pattern] === "link") {
                  // Is a link
                  outputString += "<a href='" + markdownPatterns[m].wordsInBlankSpace[1] + "'>" + markdownPatterns[m].wordsInBlankSpace[0] + "</a>";
                } else {
                  outputString += "<a class='" + markdownPatternsToClass[markdownPatterns[m].pattern] + "'>";
                }
            }
            if (markdownPatterns[m].endIndex === x) {
                outputString += "</a>";
            }
        }

        var shouldRender = true;
        for (var m = 0; m < markdownPatterns.length; m++) {
          if (markdownPatterns[m].shouldRenderChar(x, currentChar) == false) {
            shouldRender = false;
          }
        }
        if (shouldRender) {
          outputString += currentChar;
        }
    }
    console.log("Converted text to markdown")
    return outputString;
}